DNS Plugin for Infomaniak.com

Sundypha opened the pull request:
Hi @rmbloger. 
First of all: Thank you so much for this ACME client.

I added a DNS Plugin for the hoster Infomaniak from Switzerland. 

I did use the plugin on two Windows Servers and a Windows Desktop which all worked well.
I tested:
- example.com
- *.example.com
- sub.example.com
- sub2.sub1.example.com

I've got one question. Is there a way to tell the user if something failed? e.g., the zone hasn't been found? Refer to line [69](https://github.com/Sundypha/Posh-ACME/blob/dns1-infomaniak-plugin/Posh-ACME/Plugins/Infomaniak.ps1#L69) and line [150](https://github.com/Sundypha/Posh-ACME/blob/dns1-infomaniak-plugin/Posh-ACME/Plugins/Infomaniak.ps1#L150) where I could fail gracefully.

Awaiting your review and feedback!

Best
Simon

Sundypha commented:
Just figured that in the function `Find-InfomaniakZone` I query the `zone _acme-challenge.example.com` which is unnecessary. Going to create a commit shortly.

rmbolger commented:
Hi @Sundypha. Thanks for the submission! Everything looks great at first glance. Give me a bit to review and test.

rmbolger commented on Posh-ACME/Plugins/Infomaniak.ps1:
```diff
@@ -0,0 +1,256 @@
+    # - sub2.example.com
+    # - example.com
+
+    # remove _acme-challenge. from requested RecordName to get the zone
+    $zone = $RecordName.Split('.',2)[1];
```
While this might seem like an easy optimization at first, it can actually break things for people using CNAME aliases for challenges. The record being created may not always be `_acme-challenge.blah`, so you have to check the whole thing first even if most of the time it will fail.

Sundypha replied:
Thanks. Will try with an alias.

rmbolger replied:
Specifically, try running the following which is what can theoretically happen when people point an alias to the root of a domain:

```powershell
Publish-Challenge _acme-challenge.example.com (Get-PAAccount) faketoken Infomaniak $pArgs -DnsAlias aliasdomain.com -Verbose
```

In this case `aliasdomain.com` is what gets sent to the plugin in the `$RecordName` variable.

rmbolger reviewed (changes requested):
Great job. Just fix the one _acme-challenge check removal and this should be good to merge.

rmbolger commented:
> I've got one question. Is there a way to tell the user if something failed? e.g., the zone hasn't been found? Refer to line 69 and line 150 where I could fail gracefully.

Forgot to answer this. For things like the domain not being found which you can't really recover from, I'll just use `throw "my failure message"` to send an exception back up to the user so they can decide what to do.

Sundypha commented:
You were right. It failed when I had alias `abc.example.com` since it was only checking the zone `example.com` instead of the `abc` zone. Reverted the commit and added `throw`s.

rmbolger commented on Posh-ACME/Plugins/Infomaniak.ps1:
```diff
@@ -0,0 +1,262 @@
+        try {
+            $response = Invoke-RestMethod "$apiRoot/1/domain/$($zone.id)/dns/record" -Method Post -Body $json `
+                @restParams @Script:UseBasic -EA Stop
+            if($response.result -eq 'success')
+            {
+                Write-Verbose "Record $RecordName added with value $TxtValue."
+            }
+            {
+                throw "Record $RecordName with value $TxtValue could not be added."
+            }
```
I think you're missing an `else` in here.

Sundypha replied:
yes. Strange that during Publish-Challenge PS didn't throw an error.

rmbolger commented on Posh-ACME/Plugins/Infomaniak.ps1:
```diff
@@ -0,0 +1,262 @@
+        $zoneTest = $zoneTest.Split('.',2)[1]
+    }
+
+    # remove _acme-challenge. from requested zone
+    $zone = $RecordName.Split('.',2)[1]
+    Write-Debug "Zone $zone does not exist ..."
```
Another remnant of removing `_acme-challenge` to cleanup. You might just change the debug message to something like 

```powershell
Write-Debug "Zone for $RecordName not found..."
```

Sundypha replied:
Agree. Changed that for coherency.

rmbolger commented on Posh-ACME/Plugins/Infomaniak.ps1:
```diff
@@ -255,8 +256,6 @@ Function Find-InfomaniakZone {
 
-    # remove _acme-challenge. from requested zone
-    $zone = $RecordName.Split('.',2)[1]
     Write-Debug "Zone $zone does not exist ..."
```
Whoops. Now the debug message references a `$zone` variable that doesn't exist.

Sundypha replied:
Right. I'll stop programming today. Already too late. I will review the whole thing in a few hours.

rmbolger replied:
No worries. Take your time.

Sundypha commented:
Hi @rmbolger,

Did again another round of testing including timed out API calls to check debug and throw messages.
I figured now why my changes weren't persistent. VS Code opened the files at instdev.ps1 folder (\Documents\). First time writing any Powershell code.

Should I add another `$apiRoot` in the `Function Find-InfomaniakZone` for consistency?

rmbolger commented:
There should definitely be a local instance of `$apiRoot` within the function so you don't run into any unintended scoping issues. But whether you pass that instance as a parameter in the function or just hard code it doesn't really matter to me. The plugins are pretty small and having to find/replace an extra copy of the string if it ever changes is no big deal.

Sundypha commented:
I thought so exactly. I wasn't sure if Powershell has some kind of external scoping into functions called since it was working without but it felt awkward.

rmbolger commented:
Added one last fix that it looks like you had found originally and fixed in the Remove method but forgot to copy to the Add method.

Sundypha commented:
Hang on, I thought about this one. I'll try if the $RecordName can be used for adding the record instead of the short version.

Sundypha commented:
Yes, `source_idn` would work, but target_idn doesn't. The API request must have `target`.
We could change it to $RecordName, but I want to try with alias', too.
Will report shortly.

Sundypha commented:
With alias'ing the `source_idn` with `$RecordName` can't be used, so we stick with the `$recShort`.

rmbolger commented:
Thanks for the effort, @Sundypha. I'll try to get this out in a release soon.
